obfuscation_detection/utils.py

```diff
@@ -162,6 +163,42 @@ def calc_uncommon_instruction_sequences_score(function):
     score = count / sum(function_ngrams.values())
     return score
 
+arithmetic_op = set([HighLevelILOperation.HLIL_ADD, HighLevelILOperation.HLIL_NEG, HighLevelILOperation.HLIL_SUB, HighLevelILOperation.HLIL_MUL, HighLevelILOperation.HLIL_DIVS, HighLevelILOperation.HLIL_MODS])
+boolean_op = set([HighLevelILOperation.HLIL_NOT, HighLevelILOperation.HLIL_AND, HighLevelILOperation.HLIL_OR, HighLevelILOperation.HLIL_XOR])
+arithmetic_counter = 0
+boolean_counter = 0
+
+def traverse_HLIL(il):
```

Comment:
1. make the function iterative (python is bad in recursion and can hit recursion limitations for large expressions)
2. based on the heuristics logic, we can stop iterating once we have found one arithmetic and one boolean operation
3. if it is a arithmetic operation, it cannot be a boolean => if ... elif is more efficient

example:

```python
def uses_mixed_boolean_arithmetic(hlil_instruction):
    # initialize
    global ARITHMETIC_OPERATION, BOOLEAN_OPERATION
    uses_boolean = False
    uses_arithmetic = False
    todo = [hlil_instruction]

    # worklist algorithm
    while len(todo) != 0:
        instruction = todo.pop()
        # check if boolean or arithmetic operation
        if isinstance(instruction, highlevelil.HighLevelILInstruction):
            # arithmetic operation
            if instruction.operation in ARITHMETIC_OPERATION:
                uses_arithmetic = True
            # boolean operation
            elif instruction.operation in BOOLEAN_OPERATION:
                uses_boolean = True
            # mixed boolean arithmetic
            if uses_boolean and uses_arithmetic:
                return True
            # add operands to worklist
            for op in instruction.operands:
                todo.append(op)
    return False
```